ICE: `WellFormed(T)` has escaping bound vars, so it cannot be wrapped in a dummy binder.

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(non_lifetime_binders)]
#![feature(generic_const_exprs)]
#![crate_type = "lib"]

fn foo<T>(ExactSizeIterator: T) -> usize
where
    for<'a> &'a T:  ,
    for<T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
{}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.69.0-nightly (9aa5c24b7 2023-02-17)
binary: rustc
commit-hash: 9aa5c24b7d763fb98d998819571128ff2eb8a3ca
commit-date: 2023-02-17
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

### Error output

```
error[E0261]: use of undeclared lifetime name `'a`
 --> d28cc4a400eddc53a0435322d1fa5d60230fffb0.rs:8:14
  |
8 |     for<T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
  |              ^^ undeclared lifetime
  |
  = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
help: consider making the bound lifetime-generic with a new `'a` lifetime
  |
8 |     for<'a, T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
  |         +++
help: consider introducing lifetime `'a` here
  |
5 | fn foo<'a, T>(ExactSizeIterator: T) -> usize
  |        +++

error[E0261]: use of undeclared lifetime name `'a`
 --> d28cc4a400eddc53a0435322d1fa5d60230fffb0.rs:8:20
  |
8 |     for<T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
  |                    ^^ undeclared lifetime
  |
help: consider making the bound lifetime-generic with a new `'a` lifetime
  |
8 |     for<'a, T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
  |         +++
help: consider introducing lifetime `'a` here
  |
5 | fn foo<'a, T>(ExactSizeIterator: T) -> usize
  |        +++

error[E0423]: expected value, found trait `ExactSizeIterator`
 --> d28cc4a400eddc53a0435322d1fa5d60230fffb0.rs:8:60
  |
8 |     for<T> <&'a [<&'a [u32; 1] as IntoIterator>::IntoIter; ExactSizeIterator: T] as IntoIterator>::IntoIter: ExactSizeIterator,
  |                                                            ^^^^^^^^^^^^^^^^^ expecting a type here because of type ascription

warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
 --> d28cc4a400eddc53a0435322d1fa5d60230fffb0.rs:1:12
  |
1 | #![feature(non_lifetime_binders)]
  |            ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #1 <https://github.com/rust-lang/rust/issues/1> for more information
  = note: `#[warn(incomplete_features)]` on by default

warning: the feature `generic_const_exprs` is incomplete and may not be safe to use and/or cause compiler crashes
 --> d28cc4a400eddc53a0435322d1fa5d60230fffb0.rs:2:12
  |
2 | #![feature(generic_const_exprs)]
  |            ^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #76560 <https://github.com/rust-lang/rust/issues/76560> for more information
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```

thread 'rustc' panicked at '`WellFormed(T)` has escaping bound vars, so it cannot be wrapped in a dummy binder.', compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:486:13
stack backtrace:
   0:     0x7fb88dd13a5a - std::backtrace_rs::backtrace::libunwind::trace::hf4a48a91b677bc8d
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fb88dd13a5a - std::backtrace_rs::backtrace::trace_unsynchronized::h52302266f2c9e2af
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fb88dd13a5a - std::sys_common::backtrace::_print_fmt::h00aaca457620efb0
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7fb88dd13a5a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6d9f10455e422cbd
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7fb88dd739ae - core::fmt::write::h065585aa0e682316
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/core/src/fmt/mod.rs:1232:17
   5:     0x7fb88dd03e05 - std::io::Write::write_fmt::h54937c0abd9a719d
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/io/mod.rs:1684:15
   6:     0x7fb88dd13825 - std::sys_common::backtrace::_print::h0141838be2b7fc59
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7fb88dd13825 - std::sys_common::backtrace::print::h422705047959e0cf
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7fb88dd1659f - std::panicking::default_hook::{{closure}}::h04c227824228bc2c
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/panicking.rs:267:22
   9:     0x7fb88dd162db - std::panicking::default_hook::h1a17cdb165515d8b
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/panicking.rs:286:9
  10:     0x7fb88ca60224 - rustc_driver_impl[9b756e32c0d38789]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fb88dd16ddd - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h7593a5eec671dafe
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/alloc/src/boxed.rs:2002:9
  12:     0x7fb88dd16ddd - std::panicking::rust_panic_with_hook::h815681ed9ae37db1
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/panicking.rs:692:13
  13:     0x7fb88dd16b59 - std::panicking::begin_panic_handler::{{closure}}::h2a0c3307e4bad14b
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/panicking.rs:579:13
  14:     0x7fb88dd13ec6 - std::sys_common::backtrace::__rust_end_short_backtrace::hb59a12f53d5a254d
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys_common/backtrace.rs:137:18
  15:     0x7fb88dd16862 - rust_begin_unwind
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/panicking.rs:575:5
  16:     0x7fb88dd70353 - core::panicking::panic_fmt::hd1ca061751c01e2a
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/core/src/panicking.rs:64:14
  17:     0x7fb88adff682 - <rustc_hir_typeck[2a4a381332c8299]::fn_ctxt::FnCtxt>::register_wf_obligation
  18:     0x7fb88adff3e3 - <rustc_hir_typeck[2a4a381332c8299]::fn_ctxt::FnCtxt>::to_ty
  19:     0x7fb88adff27a - <rustc_hir_typeck[2a4a381332c8299]::fn_ctxt::FnCtxt>::to_ty_saving_user_provided_ty
  20:     0x7fb88adf6d8d - <rustc_hir_typeck[2a4a381332c8299]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21:     0x7fb88afb243b - <rustc_hir_typeck[2a4a381332c8299]::fn_ctxt::FnCtxt>::check_expr_coercable_to_type
  22:     0x7fb88af80a3f - <rustc_hir_typeck[2a4a381332c8299]::inherited::InheritedBuilder>::enter::<rustc_hir_typeck[2a4a381332c8299]::typeck_with_fallback<rustc_hir_typeck[2a4a381332c8299]::typeck::{closure#0}>::{closure#0}::{closure#1}, &rustc_middle[b5631d71f72482ee]::ty::typeck_results::TypeckResults>
  23:     0x7fb88af74a66 - rustc_hir_typeck[2a4a381332c8299]::typeck
  24:     0x7fb88af73163 - rustc_query_system[2b3be6912aef78d5]::query::plumbing::try_execute_query::<rustc_query_impl[a3402ef8368c6710]::queries::typeck, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt>
  25:     0x7fb88c45b9b4 - <rustc_query_impl[a3402ef8368c6710]::Queries as rustc_middle[b5631d71f72482ee]::ty::query::QueryEngine>::typeck
  26:     0x7fb88ba6be84 - rustc_mir_build[f41718743f8b2dbe]::thir::cx::thir_body
  27:     0x7fb88ba646fa - rustc_query_system[2b3be6912aef78d5]::query::plumbing::try_execute_query::<rustc_query_impl[a3402ef8368c6710]::queries::thir_body, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt>
  28:     0x7fb88c45570d - <rustc_query_impl[a3402ef8368c6710]::Queries as rustc_middle[b5631d71f72482ee]::ty::query::QueryEngine>::thir_body
  29:     0x7fb88b832aef - rustc_query_system[2b3be6912aef78d5]::query::plumbing::try_execute_query::<rustc_query_impl[a3402ef8368c6710]::queries::thir_abstract_const, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt>
  30:     0x7fb88c455ade - <rustc_query_impl[a3402ef8368c6710]::Queries as rustc_middle[b5631d71f72482ee]::ty::query::QueryEngine>::thir_abstract_const
  31:     0x7fb88ab6545e - <rustc_middle[b5631d71f72482ee]::ty::context::TyCtxt>::bound_abstract_const
  32:     0x7fb88ab650ca - <<rustc_middle[b5631d71f72482ee]::ty::context::TyCtxt>::expand_abstract_consts::Expander as rustc_type_ir[327ad7f49bc4b943]::fold::TypeFolder<rustc_middle[b5631d71f72482ee]::ty::context::TyCtxt>>::fold_const
  33:     0x7fb88ada1c27 - rustc_trait_selection[7215420893b05e21]::traits::const_evaluatable::is_const_evaluatable
  34:     0x7fb88acf3900 - <rustc_trait_selection[7215420893b05e21]::traits::fulfill::FulfillProcessor as rustc_data_structures[6722e53fadb3415d]::obligation_forest::ObligationProcessor>::process_obligation
  35:     0x7fb88aced26f - <rustc_data_structures[6722e53fadb3415d]::obligation_forest::ObligationForest<rustc_trait_selection[7215420893b05e21]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[7215420893b05e21]::traits::fulfill::FulfillProcessor>
  36:     0x7fb88b514b70 - <rustc_trait_selection[7215420893b05e21]::traits::engine::ObligationCtxt>::select_all_or_error
  37:     0x7fb88b511aa8 - rustc_hir_analysis[90a2d534158d44df]::check::wfcheck::check_item_fn
  38:     0x7fb88b50cde1 - rustc_hir_analysis[90a2d534158d44df]::check::wfcheck::check_well_formed
  39:     0x7fb88b1ac415 - rustc_query_system[2b3be6912aef78d5]::query::plumbing::get_query::<rustc_query_impl[a3402ef8368c6710]::queries::check_well_formed, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt, rustc_middle[b5631d71f72482ee]::dep_graph::dep_node::DepKind>
  40:     0x7fb88be27ba6 - rustc_data_structures[6722e53fadb3415d]::sync::par_for_each_in::<&[rustc_hir[c7d6b754e828c199]::hir::ItemId], <rustc_middle[b5631d71f72482ee]::hir::ModuleItems>::par_items<rustc_hir_analysis[90a2d534158d44df]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>
  41:     0x7fb88be27860 - rustc_hir_analysis[90a2d534158d44df]::check::wfcheck::check_mod_type_wf
  42:     0x7fb88c01971f - rustc_query_system[2b3be6912aef78d5]::query::plumbing::try_execute_query::<rustc_query_impl[a3402ef8368c6710]::queries::check_mod_type_wf, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt>
  43:     0x7fb88c0192fd - <rustc_query_impl[a3402ef8368c6710]::Queries as rustc_middle[b5631d71f72482ee]::ty::query::QueryEngine>::check_mod_type_wf
  44:     0x7fb88bd70224 - std[92d2792449a90a80]::panicking::try::<(), core[d0f7fc72c0be1fb4]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[6722e53fadb3415d]::sync::par_for_each_in<&[rustc_hir[c7d6b754e828c199]::hir_id::OwnerId], <rustc_middle[b5631d71f72482ee]::hir::map::Map>::par_for_each_module<rustc_hir_analysis[90a2d534158d44df]::check_crate::{closure#5}::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}>>
  45:     0x7fb88af080f8 - rustc_data_structures[6722e53fadb3415d]::sync::par_for_each_in::<&[rustc_hir[c7d6b754e828c199]::hir_id::OwnerId], <rustc_middle[b5631d71f72482ee]::hir::map::Map>::par_for_each_module<rustc_hir_analysis[90a2d534158d44df]::check_crate::{closure#5}::{closure#0}::{closure#0}>::{closure#0}>
  46:     0x7fb88af07f80 - <rustc_session[1c603fdce436292a]::session::Session>::track_errors::<rustc_hir_analysis[90a2d534158d44df]::check_crate::{closure#5}, ()>
  47:     0x7fb88af05258 - rustc_hir_analysis[90a2d534158d44df]::check_crate
  48:     0x7fb88aefad72 - rustc_interface[828b714f964ce1da]::passes::analysis
  49:     0x7fb88c24d4f2 - rustc_query_system[2b3be6912aef78d5]::query::plumbing::try_execute_query::<rustc_query_impl[a3402ef8368c6710]::queries::analysis, rustc_query_impl[a3402ef8368c6710]::plumbing::QueryCtxt>
  50:     0x7fb88c4550fa - <rustc_query_impl[a3402ef8368c6710]::Queries as rustc_middle[b5631d71f72482ee]::ty::query::QueryEngine>::analysis
  51:     0x7fb88bc75a23 - <rustc_interface[828b714f964ce1da]::queries::QueryResult<&rustc_middle[b5631d71f72482ee]::ty::context::GlobalCtxt>>::enter::<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}::{closure#2}::{closure#3}>
  52:     0x7fb88bc74bf4 - <rustc_interface[828b714f964ce1da]::interface::Compiler>::enter::<rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}::{closure#2}, core[d0f7fc72c0be1fb4]::result::Result<core[d0f7fc72c0be1fb4]::option::Option<rustc_interface[828b714f964ce1da]::queries::Linker>, rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>
  53:     0x7fb88bc6fe14 - rustc_span[65551fbbac8b0f76]::with_source_map::<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_interface[828b714f964ce1da]::interface::run_compiler<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  54:     0x7fb88bc6f924 - <scoped_tls[18b6b4dd1216e2e8]::ScopedKey<rustc_span[65551fbbac8b0f76]::SessionGlobals>>::set::<rustc_interface[828b714f964ce1da]::interface::run_compiler<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}>::{closure#0}, core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>
  55:     0x7fb88bc6f022 - std[92d2792449a90a80]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[828b714f964ce1da]::util::run_in_thread_pool_with_globals<rustc_interface[828b714f964ce1da]::interface::run_compiler<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}>::{closure#0}, core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>
  56:     0x7fb88bc6edcc - <<std[92d2792449a90a80]::thread::Builder>::spawn_unchecked_<rustc_interface[828b714f964ce1da]::util::run_in_thread_pool_with_globals<rustc_interface[828b714f964ce1da]::interface::run_compiler<core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>, rustc_driver_impl[9b756e32c0d38789]::run_compiler::{closure#1}>::{closure#0}, core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d0f7fc72c0be1fb4]::result::Result<(), rustc_errors[c2c7aee1e156d1f6]::ErrorGuaranteed>>::{closure#1} as core[d0f7fc72c0be1fb4]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  57:     0x7fb88dd20c63 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h24b1cedbd344db9f
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/alloc/src/boxed.rs:1988:9
  58:     0x7fb88dd20c63 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb7ac28b0dfacf269
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/alloc/src/boxed.rs:1988:9
  59:     0x7fb88dd20c63 - std::sys::unix::thread::Thread::new::thread_start::h67c8d237baa6ad51
                               at /rustc/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/library/std/src/sys/unix/thread.rs:108:17
  60:     0x7fb88968fbb5 - <unknown>
  61:     0x7fb889711d90 - <unknown>
  62:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.69.0-nightly (9aa5c24b7 2023-02-17) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `foo::{constant#1}`
#1 [thir_body] building THIR for `foo::{constant#1}`
#2 [thir_abstract_const] building an abstract representation for `foo::{constant#1}`
#3 [check_well_formed] checking that `foo` is well-formed
#4 [check_mod_type_wf] checking that types are well-formed in top-level module
#5 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0261, E0423.
For more information about an error, try `rustc --explain E0261`.
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/locales/en-US.ftl
+++ b/compiler/rustc_hir_analysis/locales/en-US.ftl
@@ -147,3 +147,11 @@ hir_analysis_main_function_generic_parameters = `main` function is not allowed t
 
 hir_analysis_variadic_function_compatible_convention = C-variadic function must have a compatible calling convention, like {$conventions}
     .label = C-variadic function must have a compatible calling convention
+
+hir_analysis_cannot_capture_late_bound_ty_in_anon_const =
+    cannot capture late-bound type parameter in a constant
+    .label = parameter defined here
+
+hir_analysis_cannot_capture_late_bound_const_in_anon_const =
+    cannot capture late-bound const parameter in a constant
+    .label = parameter defined here

--- a/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
+++ b/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
@@ -24,6 +24,8 @@ use rustc_span::symbol::{sym, Ident};
 use rustc_span::Span;
 use std::fmt;
 
+use crate::errors;
+
 trait RegionExt {
     fn early(param: &GenericParam<'_>) -> (LocalDefId, ResolvedArg);
 
@@ -161,6 +163,15 @@ enum Scope<'a> {
         s: ScopeRef<'a>,
     },
 
+    /// Disallows capturing non-lifetime binders from parent scopes.
+    ///
+    /// This is necessary for something like `for<T> [(); { /* references T */ }]:`,
+    /// since we don't do something more correct like replacing any captured
+    /// late-bound vars with early-bound params in the const's own generics.
+    AnonConstBoundary {
+        s: ScopeRef<'a>,
+    },
+
     Root {
         opt_parent_item: Option<LocalDefId>,
     },
@@ -211,6 +222,7 @@ impl<'a> fmt::Debug for TruncatedScopeDebug<'a> {
                 .field("s", &"..")
                 .finish(),
             Scope::TraitRefBoundary { s: _ } => f.debug_struct("TraitRefBoundary").finish(),
+            Scope::AnonConstBoundary { s: _ } => f.debug_struct("AnonConstBoundary").finish(),
             Scope::Root { opt_parent_item } => {
                 f.debug_struct("Root").field("opt_parent_item", &opt_parent_item).finish()
             }
@@ -312,7 +324,9 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                     break (vec![], BinderScopeType::Normal);
                 }
 
-                Scope::Elision { s, .. } | Scope::ObjectLifetimeDefault { s, .. } => {
+                Scope::Elision { s, .. }
+                | Scope::ObjectLifetimeDefault { s, .. }
+                | Scope::AnonConstBoundary { s } => {
                     scope = s;
                 }
 
@@ -1029,6 +1043,12 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
     fn visit_poly_trait_ref(&mut self, trait_ref: &'tcx hir::PolyTraitRef<'tcx>) {
         self.visit_poly_trait_ref_inner(trait_ref, NonLifetimeBinderAllowed::Allow);
     }
+
+    fn visit_anon_const(&mut self, c: &'tcx hir::AnonConst) {
+        self.with(Scope::AnonConstBoundary { s: self.scope }, |this| {
+            intravisit::walk_anon_const(this, c);
+        });
+    }
 }
 
 fn object_lifetime_default(tcx: TyCtxt<'_>, param_def_id: DefId) -> ObjectLifetimeDefault {
@@ -1267,7 +1287,8 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                 Scope::Elision { s, .. }
                 | Scope::ObjectLifetimeDefault { s, .. }
                 | Scope::Supertrait { s, .. }
-                | Scope::TraitRefBoundary { s, .. } => {
+                | Scope::TraitRefBoundary { s, .. }
+                | Scope::AnonConstBoundary { s } => {
                     scope = s;
                 }
             }
@@ -1332,7 +1353,8 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                 | Scope::Elision { s, .. }
                 | Scope::ObjectLifetimeDefault { s, .. }
                 | Scope::Supertrait { s, .. }
-                | Scope::TraitRefBoundary { s, .. } => {
+                | Scope::TraitRefBoundary { s, .. }
+                | Scope::AnonConstBoundary { s } => {
                     scope = s;
                 }
             }
@@ -1351,6 +1373,7 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
         // search.
         let mut late_depth = 0;
         let mut scope = self.scope;
+        let mut crossed_anon_const = false;
         let result = loop {
             match *scope {
                 Scope::Body { s, .. } => {
@@ -1384,10 +1407,36 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                 | Scope::TraitRefBoundary { s, .. } => {
                     scope = s;
                 }
+
+                Scope::AnonConstBoundary { s } => {
+                    crossed_anon_const = true;
+                    scope = s;
+                }
             }
         };
 
         if let Some(def) = result {
+            if let ResolvedArg::LateBound(..) = def && crossed_anon_const {
+                let use_span = self.tcx.hir().span(hir_id);
+                let def_span = self.tcx.def_span(param_def_id);
+                match self.tcx.def_kind(param_def_id) {
+                    DefKind::ConstParam => {
+                        self.tcx.sess.emit_err(errors::CannotCaptureLateBoundInAnonConst::Const {
+                            use_span,
+                            def_span,
+                        });
+                    }
+                    DefKind::TyParam => {
+                        self.tcx.sess.emit_err(errors::CannotCaptureLateBoundInAnonConst::Type {
+                            use_span,
+                            def_span,
+                        });
+                    }
+                    _ => unreachable!(),
+                }
+                return;
+            }
+
             self.map.defs.insert(hir_id, def);
             return;
         }
@@ -1465,7 +1514,8 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                         | Scope::Elision { s, .. }
                         | Scope::ObjectLifetimeDefault { s, .. }
                         | Scope::Supertrait { s, .. }
-                        | Scope::TraitRefBoundary { s, .. } => {
+                        | Scope::TraitRefBoundary { s, .. }
+                        | Scope::AnonConstBoundary { s } => {
                             scope = s;
                         }
                     }
@@ -1701,7 +1751,9 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
 
                 Scope::ObjectLifetimeDefault { lifetime: Some(l), .. } => break l,
 
-                Scope::Supertrait { s, .. } | Scope::TraitRefBoundary { s, .. } => {
+                Scope::Supertrait { s, .. }
+                | Scope::TraitRefBoundary { s, .. }
+                | Scope::AnonConstBoundary { s } => {
                     scope = s;
                 }
             }

--- a/compiler/rustc_hir_analysis/src/errors.rs
+++ b/compiler/rustc_hir_analysis/src/errors.rs
@@ -381,3 +381,21 @@ pub(crate) struct VariadicFunctionCompatibleConvention<'a> {
     pub span: Span,
     pub conventions: &'a str,
 }
+
+#[derive(Diagnostic)]
+pub(crate) enum CannotCaptureLateBoundInAnonConst {
+    #[diag(hir_analysis_cannot_capture_late_bound_ty_in_anon_const)]
+    Type {
+        #[primary_span]
+        use_span: Span,
+        #[label]
+        def_span: Span,
+    },
+    #[diag(hir_analysis_cannot_capture_late_bound_const_in_anon_const)]
+    Const {
+        #[primary_span]
+        use_span: Span,
+        #[label]
+        def_span: Span,
+    },
+}

--- a/tests/ui/traits/non_lifetime_binders/late-bound-in-anon-ct.rs
+++ b/tests/ui/traits/non_lifetime_binders/late-bound-in-anon-ct.rs
@@ -1,0 +1,11 @@
+#![feature(non_lifetime_binders, generic_const_exprs)]
+//~^ WARN the feature `non_lifetime_binders` is incomplete
+//~| WARN the feature `generic_const_exprs` is incomplete
+
+fn foo() -> usize
+where
+    for<T> [i32; { let _: T = todo!(); 0 }]:,
+    //~^ ERROR cannot capture late-bound type parameter in a constant
+{}
+
+fn main() {}

--- a/tests/ui/traits/non_lifetime_binders/late-bound-in-anon-ct.stderr
+++ b/tests/ui/traits/non_lifetime_binders/late-bound-in-anon-ct.stderr
@@ -1,0 +1,27 @@
+warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/late-bound-in-anon-ct.rs:1:12
+   |
+LL | #![feature(non_lifetime_binders, generic_const_exprs)]
+   |            ^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
+   = note: `#[warn(incomplete_features)]` on by default
+
+warning: the feature `generic_const_exprs` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/late-bound-in-anon-ct.rs:1:34
+   |
+LL | #![feature(non_lifetime_binders, generic_const_exprs)]
+   |                                  ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #76560 <https://github.com/rust-lang/rust/issues/76560> for more information
+
+error: cannot capture late-bound type parameter in a constant
+  --> $DIR/late-bound-in-anon-ct.rs:7:27
+   |
+LL |     for<T> [i32; { let _: T = todo!(); 0 }]:,
+   |         -                 ^
+   |         |
+   |         parameter defined here
+
+error: aborting due to previous error; 2 warnings emitted
+

